server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -799,133 +803,174 @@ public static class Fields {
     static final String METADATA_FIELDS = "metadata_fields";
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
         assert hasReferences();
-        builder.startObject();
-        toInnerXContent(builder, params);
-        builder.endObject();
-        return builder;
+        return Iterators.concat(
+            ChunkedToXContentHelper.startObject(),
+            // Iterators.flatMap(toInnerXContentChunked(params), c -> c.toXContent(params)),
+            toInnerXContentChunked(params),
+            ChunkedToXContentHelper.endObject()
+        );
+    }
+
+    private static <T> Iterator<ToXContent> chunkedSection(
+        ToXContent before,
+        Iterator<T> items,
+        Function<T, Iterator<ToXContent>> fn,
+        ToXContent after
+    ) {
+        return Iterators.concat(Iterators.single(before), Iterators.flatMap(items, fn::apply), Iterators.single(after));
     }
 
     // public because we render hit as part of completion suggestion option
-    public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) throws IOException {
+    public Iterator<? extends ToXContent> toInnerXContentChunked(ToXContent.Params outerParams) {
+        var chunks = new ArrayList<Iterator<? extends ToXContent>>();
```

Comment:
Should we define the size here to avoid having to grow & copy the array when it exceeds the default initial size of 10?

Reply:
Good point, we already know the max number of "content generators" that will go into it, so we can size it accordingly.